zombsole/game.py

```diff
@@ -216,22 +222,45 @@ def spawn_zombies_to_maintain_minimum(self):
     def encode_position_as_channels(self, position):
         """Get the character to draw for a given position of the world."""
         # decorations first, then things over them
-        thing = (self.world.things.get(position) or
-                 self.world.decoration.get(position))
+        if self.map.within_bounds(position):
+            thing = (self.world.things.get(position) or
+                     self.world.decoration.get(position))
+        else:
+            thing = Wall(position) # Note the position is out of bounds here
 
         if thing is not None:
             life = getattr(thing, 'life', 0)
             weapon = getattr(thing, 'weapon', None)
             weapon_name = weapon.name if weapon is not None else 'none'
             weapon_code = self.weapon_labels.get(weapon_name, 0)
+            thing_code = self.thing_labels.get(thing.icon_basic, 0)
+            if thing_code == 7: # agent
+                thing_code = 8 + int(thing.agent_id)
             return [
-                ord(thing.icon_basic),
+                thing_code,
                 life,
                 weapon_code
             ]
         else:
             return [0, 0, 0]
     
+    def encode_world_with_channels(self):
+        """Render the world using channels."""
+        return [
+            [self.encode_position_as_channels((x, y)) for x in range(self.world.size[0])]
+            for y in range(self.world.size[1])
+        ]
+    
+    def encode_surroundings_with_channels(self, position: Tuple[int, int], surroundings_half_width: int):
+        """Render the surroundings using channels."""
+        xrange = range(position[0] - surroundings_half_width, position[0] + surroundings_half_width + 1)
+        yrange = range(position[1] - surroundings_half_width, position[1] + surroundings_half_width + 1)
+
+        return [
+            [self.encode_position_as_channels((x, y)) for x in xrange]
+            for y in yrange
+        ]
+    
     def encode_position_simple(self, position):
         """Get the character to draw for a given position of the world."""
         # decorations first, then things over them
```

Comment:
Might need a guard for an out-of-bounds position, as with `encode_position_with_channels`.